File: frontend/src/types/activity.ts
import type { PaginationParams } from "@/types/pagination";

export const activityTypes = ["call", "email", "meeting", "follow_up"] as const;

export type ActivityType = (typeof activityTypes)[number];

export type Activity = {
  company_id: string | null;
  completed: boolean;
  contact_id: string | null;
  created_at: string;
  description: string | null;
  due_date: string | null;
  id: string;
  lead_id: string | null;
  organization_id: string;
  title: string;
  type: ActivityType;
  user_id: string;
};

export type ActivityInput = {
  company_id: string | null;
  completed: boolean;
  contact_id: string | null;
  description: string | null;
  due_date: string | null;
  lead_id: string | null;
  title: string;
  type: ActivityType;
  user_id?: string | null;
};

export type ActivitiesListParams = PaginationParams & {
  completed?: boolean;
  company_id?: string;
  contact_id?: string;
  lead_id?: string;
  sort_by?: "created_at" | "due_date" | "type";
  sort_direction?: "asc" | "desc";
  type?: ActivityType;
  user_id?: string;
};

File: backend/app/schemas/activity.py
from datetime import datetime
from typing import Annotated, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, StringConstraints, field_validator

from app.schemas.common import RequestSchema

ActivityType = Literal["call", "email", "meeting", "follow_up"]
ActivityTitle = Annotated[str, StringConstraints(min_length=1, max_length=255)]
ActivityDescription = Annotated[str, StringConstraints(max_length=20_000)]


class ActivityCreate(RequestSchema):
    type: ActivityType
    title: ActivityTitle
    description: ActivityDescription | None = None
    due_date: datetime | None = None
    completed: bool = False
    user_id: UUID | None = None
    company_id: UUID | None = None
    contact_id: UUID | None = None
    lead_id: UUID | None = None


class ActivityUpdate(RequestSchema):
    type: ActivityType | None = None
    title: ActivityTitle | None = None
    description: ActivityDescription | None = None
    due_date: datetime | None = None
    completed: bool | None = None
    user_id: UUID | None = None
    company_id: UUID | None = None
    contact_id: UUID | None = None
    lead_id: UUID | None = None

    @field_validator("type", "title", "user_id", "completed")
    @classmethod
    def require_non_null_required_fields(cls, value: object) -> object:
        if value is None:
            raise ValueError("This field cannot be null")
        return value


class ActivityResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    organization_id: UUID
    user_id: UUID
    company_id: UUID | None
    contact_id: UUID | None
    lead_id: UUID | None
    type: ActivityType
    title: str
    description: str | None
    due_date: datetime | None
    completed: bool
    created_at: datetime
